source.go

```diff
@@ -574,3 +570,45 @@ func (s *Spec) getPatchedSources(sOpt SourceOpts, worker llb.State, filterFunc f
 
 	return PatchSources(worker, s, states, opts...), nil
 }
+
+// Tar creates a tar+gz from the provided state and puts it in the provided dest.
+// The provided work state is used to perform the necessary operations to produce the tarball and requires the tar and gzip binaries.
+func Tar(work, src llb.State, dest string, opts ...llb.ConstraintsOpt) llb.State {
+	// Put the output tar in a consistent location regardless of `dest`
+	// This way if `dest` changes we don't have to rebuild the tarball, which can be expensive.
+	outBase := "/tmp/out"
+	out := filepath.Join(outBase, filepath.Dir(dest))
+	worker := work.Run(
+		llb.AddMount("/src", src, llb.Readonly),
+		llb.Args([]string{"/bin/sh", "-c", "tar -C /src -cvzf /tmp/st ."}),
```

Comment:
q: could this use the helper function `ShArgs`? same question for `587`

Reply:
Done.